### PR TITLE
spring-with-qualified-op-name-scenario + spring-with-request-method-op-name-scenario

### PR DESCRIPTION
**Pull request by BFergerson**
I'm running into an issue when trying to verify this. I'm getting:
```
start docker container
start docker container
Creating network "spring-with-qualified-op-name-scenario-spring-resttemplate-402_annotation-402_restannotation-500_default" with the default driver
Creating network "spring-with-qualified-op-name-scenario-spring-resttemplate-401_annotation-401_restannotation-501_default" with the default driver
Pulling spring-with-qualified-op-name-scenario (skywalking/spring-with-qualified-op-name-scenario:Spring-RestTemplate-4.0.2_Annotation-4.0.2_RestAnnotation-5.0.0)...
Pulling spring-with-qualified-op-name-scenario (skywalking/spring-with-qualified-op-name-scenario:Spring-RestTemplate-4.0.1_Annotation-4.0.1_RestAnnotation-5.0.1)...
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]ERROR: pull access denied for skywalking/spring-with-qualified-op-name-scenario, repository does not exist or may require 'docker login'
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]ERROR: pull access denied for skywalking/spring-with-qualified-op-name-scenario, repository does not exist or may require 'docker login'
3 containers are running, The validate program will be retried in 40 seconds.
http://localhost:18180/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case
http://localhost:18080/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case
```

I can't access the `http://localhost:18180/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case` url so I'm not able to make the correct `expectedData.yaml` file.

Any clue what I'm doing wrong?

### CODE REVIEW

**Comment by wu-sheng** [comment]
@ascrutae Please verify the test case PR.

File: spring-with-qualified-op-name-scenario/pom.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>com.spotify</groupId>
+                <artifactId>docker-maven-plugin</artifactId>
+                <version>0.4.13</version>
+                <configuration>
+                    <imageName>skywalking/spring-with-qualified-op-name-scenario</imageName>
+                    <imageTags>
+                        <imageTag>1.0.0</imageTag>
```
Comment (ascrutae) [inline spring-with-qualified-op-name-scenario/pom.xml]: the version of imageTag should be `${docker.image.version}`

**Comment by BFergerson** [comment]
@ascrutae, thank you. I've implemented the `expectedData.yml` file and even though when I do a diff on the expected vs actual data I'm getting the following validation issues:
```
tags: 
 expected: http://localhost:8080/spring-with-qualified-op-name-scenario/qualifiedname/acceptAllExplicit
 actual: http://localhost:8080/spring-with-qualified-op-name-scenario/qualifiedname/acceptAllImplicit
tags: 
 expected: http://localhost:18180/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case
 actual: http://localhost:18080/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case
[2019-03-16 16:48:20:360] [ERROR] - org.skywalking.apm.test.agent.tool.Main.main(Main.java:40) - assert failed.
assert application[spring-with-qualified-op-name-scenario] segment: 
 expected count : 1 
 actual: 0
```

When I do a diff on the expected vs actual the only differences are on fields:
 - segmentId
 - parentTraceSegmentId
 - startTime
 - endTime

I believe all of those are able to be different (+ parentTraceSgementId which is set dynamically). How does it think the tags are wrong if even in the actual data vs expected data it shows it to be correct?

Also, in my `expectedData.yml` I'm using `http://localhost:18180` for the expected url. I don't see any other tests using those type of ports so I imagine that's wrong. What should I be using given that it is expecting those ports?

File: spring-with-qualified-op-name-scenario/config/expectedData.yaml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      spanType: Exit
+      peer: "localhost:8080"
+      peerId: 0
+    - operationName: "test.apache.skywalking.testcase.qualifiedname.controller.SpringEndpointQualifiedNameController.qualifiedNameCase()"
+      operationId: 0
+      parentSpanId: -1
+      spanId: 0
+      spanLayer: Http
+      tags:
+      - {key: "url", value: "http://localhost:18180/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case"}
```
Comment (ascrutae) [inline spring-with-qualified-op-name-scenario/config/expectedData.yaml]:
```suggestion
      - {key: "url", value: "http://localhost:{SERVER_OUTPUT_PORT}/spring-with-qualified-op-name-scenario/qualifiedname/qualifiedname-case"}
```

**Review by ascrutae — DISMISSED** [review]
LGTM

**Comment by BFergerson** [comment]
@ascrutae, thanks again. Last issue. It's saying there is a validation error on the tags. Like so:
```
tags: 
 expected: http://localhost:8080/spring-with-qualified-op-name-scenario/qualifiedname/acceptAllExplicit
 actual: http://localhost:8080/spring-with-qualified-op-name-scenario/qualifiedname/acceptAllImplicit
```

Again, if I do a diff on the actual vs expected there is no difference besides the expected fields and it doesn't tell me in which segment/span the validation issue is occurring in. I've quadruple checked that all `GET` requests are going to `acceptAllImplicit` and all `POST` requests are going to `acceptAllExplicit`. This is how the integration test code works and how the `expectedData.yml` file expects. So what am I doing wrong with the tags?

**Comment by ascrutae** [comment]
The port is different from the port your export in `docker-compose.yml`, because of the autotest program runs multiple test cases at the same time, and the port need replace with the real port in run time.

**Comment by ascrutae** [comment]
Because no one knows the upload sequence of the trace segment, so the integration test needs to compare all segments in `expecatedData.yml,` that's is the warning message you saw. You should read all assert message if you want to know the real reason why the test case failed.

**Review by wu-sheng — CHANGES_REQUESTED** [review]
`test_framework` and `support-versions` names in `testcase.yml` need to be changed, based on case names. Also profile name in `pom.xml` need to match those.

**Comment by BFergerson** [comment]
@wu-sheng, I wasn't sure what those values were supposed to be so I copied the `spring-scenario`. Instead of `Spring-RestTemplate_Annotation_RestAnnotation` what are you suggesting? I thought these values were what I needed to test multiple versions of Spring (as you initially suggested)?

**Comment by wu-sheng** [comment]
You could click the report links in your main repo pr. You will see those names. I am sorry, we don't make the meaning clear.

Case name shows as a group in test report, and versions just match version list in each group.

**Comment by wu-sheng** [comment]
You will see, there are two comments by robot, our test robot. They look like same, but actually they are based on the two different cases, you provide in this pr.

**Comment by wu-sheng** [comment]
Name js not quite important to be very accurate, but need to keep them clear and unique.

**Comment by BFergerson** [comment]
@wu-sheng, sorry but I'm not sure what you're asking me to change it to. I have two test scenarios: `spring-with-qualified-op-name-scenario` and `spring-with-request-method-op-name-scenario`. I can modify the `test_framework` for both of them but I'm not sure what kind of values you're looking for. I'm still using Spring and the REST annotations. It seems to fit the scenarios I added. If it doesn't, could you give me an example of something that would?

**Comment by wu-sheng** [comment]
But you can see there are two test results
- https://github.com/apache/incubator-skywalking/pull/2353#issuecomment-473671686
- https://github.com/apache/incubator-skywalking/pull/2353#issuecomment-473672630

Each of them is for one of your case, but you can see, there looks like the same. The reason is the following two settings are same.
- [`test_framework`](https://github.com/BFergerson/agent-auto-integration-testcases/blob/master/spring-with-request-method-op-name-scenario/testcase.yml#L14)
- [`support_versions `](https://github.com/BFergerson/agent-auto-integration-testcases/blob/master/spring-with-qualified-op-name-scenario/testcase.yml#L22-L24)

This is the whole report for daily, https://github.com/SkyAPMTest/agent-integration-test-report. In order to make it clear to read, I suggest you use `SpringMVC` for both folder cases. And in version, we decribe the difference, such as `spring-4.0.1-with-qualified-op-name-scenario`, `spring-4.0.1-with-request-method-op-name-scenario`.

Then the tool could group these two cases(including 4 versions) together, like this, https://github.com/SkyAPMTest/agent-integration-test-report#Jedis

Does this make sense now?

@ascrutae I think your document is misguiding poeple, could you adjust it a little?

**Comment by ascrutae** [comment]
ok, Got it , I will do it

**Comment by BFergerson** [comment]
@wu-sheng, makes sense, thanks. Does everything look correct now?

**Comment by wu-sheng** [comment]
LGTM  @ascrutae please verify and run the cases

**Review by wu-sheng — APPROVED** [review]

**Review by ascrutae — APPROVED** [review]

**Pull request merged**